forgerock-openbanking-uk-aspsp-rs/forgerock-openbanking-uk-aspsp-rs-mock-store/forgerock-openbanking-uk-aspsp-rs-mock-store-server/src/main/java/com/forgerock/openbanking/aspsp/rs/store/repository/migration/v3_1_6/MongoDbPaymentsChangeLog.java

```diff
@@ -24,83 +24,215 @@
+                    }
+                }
+        );
+        // run the bulk operation
+        if (!updates.isEmpty()) {
+            docsUpdated = bulkOperations.updateMulti(updates).execute().getModifiedCount();
+        }
+        // execution time control - stop
+        stop(legacyClazz, docsUpdated, elapsedTime);
+        return docsUpdated;
     }
 
-    private void migrateInternationalPayments(MongoTemplate mongoTemplate) {
-        CloseableIterator<FRInternationalConsent2> frInternationalConsents = getLegacyDocuments(mongoTemplate, FRInternationalConsent2.class);
-        frInternationalConsents.forEachRemaining(d -> migrate(mongoTemplate, d, toFRInternationalConsent(d)));
+    /*
+     * Introspect the legacyClassDocInstance to get the id and countrySubDivision values to prepare the bulk update operation
+     * @param legacyClazz collection class
+     * @param legacyClassDocInstance document legacyClass instance
+     * @param writeRiskParentField field name to get the write object that contains the risk object related with the collection class
+     * @param <T> generic parameter to conforms legacyClazz object
+     * @return a optional pair with the id and the first value of countrySubDivision
+     * @throws IntrospectionException
+     * @throws InvocationTargetException
+     * @throws IllegalAccessException
+     */
+    private <T> Optional<Pair<String, String>> getPair(Class<T> legacyClazz, Object legacyClassDocInstance, String writeRiskParentField) throws IntrospectionException, InvocationTargetException, IllegalAccessException {
+        Optional<Method> optionalMethod = Arrays.stream(legacyClazz.getDeclaredMethods()).filter(m -> m.getName().toLowerCase().endsWith(writeRiskParentField.toLowerCase())).findFirst();
+        if (optionalMethod.isPresent()) {
+            String division = null;
+            String id;
+            // get the id value
+            PropertyDescriptor pd = new PropertyDescriptor("id", legacyClassDocInstance.getClass());
+            id = (String) pd.getReadMethod().invoke(legacyClassDocInstance);
+
+            // get the write risk parent object
+            pd = new PropertyDescriptor(writeRiskParentField, legacyClassDocInstance.getClass());
+            Object obj = pd.getReadMethod().invoke(legacyClassDocInstance);
+
+            // get the risk object
+            pd = new PropertyDescriptor(RISK, obj.getClass());
+            obj = pd.getReadMethod().invoke(obj);
+            log.info(obj.toString());
+
+            // get the delivery adress object
+            pd = new PropertyDescriptor(DELIVERY_ADDRESS, obj.getClass());
+            obj = pd.getReadMethod().invoke(obj);
+            log.info(obj.toString());
+
+            // get the country sub division object
+            pd = new PropertyDescriptor(COUNTRY_SUB_DIVISION, obj.getClass());
+            obj = pd.getReadMethod().invoke(obj);
```

Comment:
To be honest, I'm a bit lost in all of this. I'm wondering if it would be easier if each legacy document implemented an interface like this?

```
/**
 * Represents the ID and country subdivision fields within each legacy document type. This allows us to treat each
 * document in a type safe generic way.
 */
@Deprecated
public interface LegacySubDivisionDocument {

    String getId();

    String getCountrySubDivision();
}
```

Then each document could have something like:
```
@Override
public String getCountrySubDivision() {
   if (domesticConsent.getRisk() == null || domesticConsent.getRisk().getDeliveryAddress() == null) {
      return null;
   }
   FRPaymentRisk.FRRiskDeliveryAddress deliveryAddress = domesticConsent.getRisk().getDeliveryAddress();
   return deliveryAddress.getCountrySubDivision() == null ? null : deliveryAddress.getCountrySubDivision().get(0);
}
```
Perhaps that would help and reduce some of this reflection?

Reply:
Done